contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -223,24 +223,96 @@ contract MixinExchangeWrapper {
             // Account for the ERC20Bridge transfering more of the maker asset than expected.
             if (makerAssetProxyId == erc20BridgeProxyId) {
                 uint256 balanceAfter = IERC20Token(tokenAddress).balanceOf(address(this));
-                makerAssetAcquiredAmount = LibSafeMath.max256(
+                sellFillResults.makerAssetAcquiredAmount = LibSafeMath.max256(
                     balanceAfter.safeSub(balanceBefore),
-                    makerAssetAcquiredAmount
+                    sellFillResults.makerAssetAcquiredAmount
                 );
             }
 
-            orders[i].makerAssetData.transferOut(makerAssetAcquiredAmount);
+            orders[i].makerAssetData.transferOut(sellFillResults.makerAssetAcquiredAmount);
 
             totalWethSpentAmount = totalWethSpentAmount
-                .safeAdd(wethSpentAmount);
+                .safeAdd(sellFillResults.wethSpentAmount)
+                .safeAdd(sellFillResults.protocolFeePaid);
             totalMakerAssetAcquiredAmount = totalMakerAssetAcquiredAmount
-                .safeAdd(makerAssetAcquiredAmount);
+                .safeAdd(sellFillResults.makerAssetAcquiredAmount);
+
+            // Stop execution if the entire amount of WETH has been sold
+            if (totalWethSpentAmount >= wethSellAmount) {
+                break;
+            }
+        }
+    }
+
+    /// @dev Synchronously executes multiple calls of fillOrder until total amount of WETH (exclusive of protocol fee)
+    ///      has been sold by taker.
+    /// @param orders Array of order specifications.
+    /// @param wethSellAmount Desired amount of WETH to sell.
+    /// @param signatures Proofs that orders have been signed by makers.
+    /// @return totalWethSpentAmount Total amount of WETH spent on the given orders.
+    /// @return totalMakerAssetAcquiredAmount Total amount of maker asset acquired from the given orders.
+    function _marketSellExactAmountNoThrow(
+        LibOrder.Order[] memory orders,
+        uint256 wethSellAmount,
+        bytes[] memory signatures
+    )
+        internal
+        returns (
+            uint256 totalWethSpentAmount,
+            uint256 totalMakerAssetAcquiredAmount
+        )
+    {
+        bytes4 erc20BridgeProxyId = IAssetData(address(0)).ERC20Bridge.selector;
+        uint256 totalProtocolFeePaid;
+
+        for (uint256 i = 0; i != orders.length; i++) {
+            // Preemptively skip to avoid division by zero in _marketSellSingleOrder
+            if (orders[i].makerAssetAmount == 0 || orders[i].takerAssetAmount == 0) {
+                continue;
+            }
+
+            // The remaining amount of WETH to sell
+            uint256 remainingTakerAssetFillAmount = wethSellAmount
+                .safeSub(totalWethSpentAmount);
+
```

Comment:
What if we also did something like this here:

```solidity
if (remainingTakerAssetFillAmount > singleProtocolFee) {
    // Do not count the protocol fee as part of the fill amount.
    remainingTakerAssetFillAmount = remainingTakerAssetFillAmount.safeSub(singleProtocolFee);
} else {
    // Stop if we don't have at least enough ETH to pay another protocol fee.
    break;
}
```

Reply:
My hunch is line 206 (`.safeSub(_isV2Order(orders[i]) ? 0 : protocolFee);`)  in `_marketSellNoThrow()` is probably where I was seeing those rigo + USDC + high gas reverts, so adding something similar to that function could also fix that.

Reply:
Perhaps the naming is off right now, but `wethSellAmount` in the parameter is exclusive of protocol fees. So we don't want to subtract the protocolFee as it is not included in `wethSellAmount`. I.e A user calls `marketSellAmountWithEth` with 2 ETH and provides 2.1 ETH in value. We then proceed to sell 2ETH, not 2.1 ETH.